### DISCUSSION TITLE
NiceGUI and replit.com

### DISCUSSION THREAD

**Discussion by CCR010** [zauberzeug/nicegui#539] [Q&A]
Can anyone help me get NiceGUI to work in replit.com? I am getting this error. Thanks. 

```
Traceback (most recent call last):
  File "main.py", line 1, in <module>
    import nicegui
  File "/home/runner/niceguitest/venv/lib/python3.10/site-packages/nicegui/__init__.py", line 1, in <module>
    from . import elements, globals, ui
  File "/home/runner/niceguitest/venv/lib/python3.10/site-packages/nicegui/ui.py", line 4, in <module>
    from .element import Element as element
  File "/home/runner/niceguitest/venv/lib/python3.10/site-packages/nicegui/element.py", line 8, in <module>
    from typing_extensions import Self
ImportError: cannot import name 'Self' from 'typing_extensions' (/home/runner/niceguitest/venv/lib/python3.10/site-packages/typing_extensions.py)
```

**Comment by falkoschindler** [comment]
Please have a look over here: https://github.com/zauberzeug/nicegui/issues/431
I think @gloomie had the same problem and found a solution.

**Reply by CCR010** [reply]
thanks @falkoschindler this looks to have resolved the typing_extensions issue. Now when running i am getting these errors..
```
NiceGUI ready to go on http://0.0.0.0:8080
http://45a713f4-d656-447a-959a-0e888cec4eab.id.repl.co/_nicegui/static/quasar.umd.prod.js not found
http://45a713f4-d656-447a-959a-0e888cec4eab.id.repl.co/_nicegui/static/nicegui.css not found
http://45a713f4-d656-447a-959a-0e888cec4eab.id.repl.co/_nicegui/static/fonts.css not found
http://45a713f4-d656-447a-959a-0e888cec4eab.id.repl.co/_nicegui/static/quasar.prod.css not found
http://45a713f4-d656-447a-959a-0e888cec4eab.id.repl.co/_nicegui/static/socket.io.min.js not found
http://45a713f4-d656-447a-959a-0e888cec4eab.id.repl.co/_nicegui/static/vue.global.prod.js not found
http://45a713f4-d656-447a-959a-0e888cec4eab.id.repl.co/_nicegui/static/tailwind.min.js not found
```

**Reply by falkoschindler** [reply]
Just for the record: Did `pip install --upgrade typing_extensions` solve the problem with this module? I just tried it myself and this command seems to fix it.
The "not found" errors remain, though.

**Comment by gloomie** [comment]
Unfortunately i didnt find the solution to this...
The problem occurs not only on replit but on my desktop too. I dont know why.

**Comment by rodja** [comment]
It might be caused by an outdated version of the 'typing_extensions' module.
Try updating the 'typing_extensions' module to the latest version:

    pip install --upgrade typing_extensions

**Reply by falkoschindler** [reply]
@rodja Like in https://github.com/zauberzeug/nicegui/issues/431 there are two problems. The issue with installing the right typing_extensions package [can be solved](https://github.com/zauberzeug/nicegui/issues/431#issuecomment-1444085593). But then there is the routing problem, for which we have no solution yet.

**Reply by rodja** [reply]
Oh yes. I missed that. Thanks. #431 is referring to https://github.com/zauberzeug/nicegui/issues/386#issuecomment-1434141048. I'll repeat the idea below for clarity.

**Comment by rodja** [comment]
As written in https://github.com/zauberzeug/nicegui/issues/386#issuecomment-1434141048 we can maybe solve the routing problem by manipulating the `prefix` used to access NiceGUI ressources:

https://github.com/zauberzeug/nicegui/blob/57d6ad884457cc78e58d1457f3a5a6dad343a654/nicegui/client.py#L68

We use this to support reverse-proxy on sub paths (see #245 and #188). These proxies set the `X-Forwarded-Prefix` header and needs to be taken into account when serving the page. Maybe you can try with a locally modified version of NiceGUI to set the prefix to `cgi-bin/nicegui.cgi`? If that works we can add a configuration option to set the prefix in `ui.run` or as environmental variable.

**Reply by falkoschindler** [reply]
I just tried it. In the "Files" panel you can show hidden files (use the three dots) and navigate to venv/lib/python3.10/site-packages/nicegui/client.py. Then I added the line 74 `prefix = 'cgi-bin/nicegui.cgi'`.

But the problem remains:

```
http://d7e14fc9-e1c8-42eb-b86f-d31ae583e283.id.repl.co/cgi-bin/nicegui.cgi/_nicegui/static/socket.io.min.js not found
http://d7e14fc9-e1c8-42eb-b86f-d31ae583e283.id.repl.co/cgi-bin/nicegui.cgi/_nicegui/static/vue.global.prod.js not found
http://d7e14fc9-e1c8-42eb-b86f-d31ae583e283.id.repl.co/cgi-bin/nicegui.cgi/_nicegui/static/fonts.css not found
http://d7e14fc9-e1c8-42eb-b86f-d31ae583e283.id.repl.co/cgi-bin/nicegui.cgi/_nicegui/static/nicegui.css not found
http://d7e14fc9-e1c8-42eb-b86f-d31ae583e283.id.repl.co/cgi-bin/nicegui.cgi/_nicegui/static/quasar.umd.prod.js not found
http://d7e14fc9-e1c8-42eb-b86f-d31ae583e283.id.repl.co/cgi-bin/nicegui.cgi/_nicegui/static/quasar.prod.css not found
http://d7e14fc9-e1c8-42eb-b86f-d31ae583e283.id.repl.co/cgi-bin/nicegui.cgi/_nicegui/static/tailwind.min.js not found
```

**Reply by falkoschindler** [reply]
I think "cgi-bin/nicegui.cgi" is a custom path for #386. But there's no reason this should work on replit.com.

**Comment by falkoschindler** [comment]
Very interesting: Most routes are working. Only files from the static folder return a 404.
![Screenshot 2023-03-17 at 09 54 29](https://user-images.githubusercontent.com/5767091/225858073-901057f6-2cea-4576-a74c-25ea478d7d0e.png)

**Reply by falkoschindler** [reply]
There's a difference in how they are mounted:

nicegui.py
```py
app.mount('/_nicegui/static', StaticFiles(directory=Path(__file__).parent / 'static'), name='static')

...

@app.get('/_nicegui/dependencies/{id}/{name}')
def get_dependencies(id: int, name: str):
    if id in js_dependencies and js_dependencies[id].path.exists() and js_dependencies[id].path.name == name:
        return FileResponse(js_dependencies[id].path, media_type='text/javascript')
    raise HTTPException(status_code=404, detail=f'dependency "{name}" with ID {id} not found')
```

Apparently app.mount does not work as expected on replit.com.

**Reply by falkoschindler** [reply]
Replacing `app.mount` with a custom endpoint "solves" the issue (apart from fronts which live one level deeper):

```py
@app.get('/_nicegui/static/{name}')
def get_static(name: str):
    return FileResponse(Path(__file__).parent / 'static' / name)
```

But I'd like to find the root cause.

**Reply by falkoschindler** [reply]
In a simple FastAPI app both routing methods work just fine:
https://replit.com/@FalkoSchindler/FastAPI-Starter

**Comment by falkoschindler** [comment]
I found something:
Mounting static files works when we set `StaticFiles(..., follow_symlink=True)`. This way we can fix NiceGUI on Replit.com. 😏

**Comment by falkoschindler** [comment]
Since version 1.2.5 NiceGUI should be working on replit.com. Feel free to re-open the discussion if there're still problems.

**Reply by ndemou** [reply]
I just tested and it didn't succeed at first (same "ImportError: cannot import name 'Self'")

Tried `pip install --upgrade typing_extensions` with no success either (same complains about incompatible typing_extensions versions) 

Then out of desperation I restarted my replit and it just worked... 🤷 Noting it here for others
